jb/fftw/time_delay_estimator_many.hpp

```diff
@@ -0,0 +1,172 @@
+#ifndef jb_fftw_time_delay_estimator_many_hpp
+#define jb_fftw_time_delay_estimator_many_hpp
+
+#include <jb/fftw/aligned_multi_array.hpp>
+#include <jb/fftw/alignment_traits.hpp>
+#include <jb/fftw/plan.hpp>
+#include <jb/testing/check_complex_close_enough.hpp>
+#include <jb/complex_traits.hpp>
+
+namespace jb {
+namespace fftw {
+
+/**
+ * A simple time delay estimator based on cross-correlation.
+ *
+ * Timeseries are implemented as aligned multi arrays of N dimensions
+ *
+ * @tparam array_t timeseries multi array type
+ * @tparam T timeseries value type
+ * @tparam K timeseries dimensionality
+ */
+
+template <typename array_t, template <typename T, std::size_t K>
+                            class multi_array = jb::fftw::aligned_multi_array>
+class time_delay_estimator_many {
+public:
+  //@{
+
+  /**
+   * @name Type traits
+   */
+  /// The input timeseries type
+  using array_type = array_t;
+
+  /// The values stored in the input timeseries
+  using element_type =
+      typename jb::detail::array_traits<array_type>::element_type;
+
+  /// Extract T out of std::complex<T>, otherwise simply T.
+  using precision_type =
+      typename jb::extract_value_type<element_type>::precision;
+
+  /// The type used to store the DFT of the input timeseries
+  using frequency_array_type =
+      multi_array<std::complex<precision_type>, array_type::dimensionality>;
+
+  /// The type used to store the inverse of the DFT
+  using output_array_type =
+      multi_array<precision_type, array_type::dimensionality>;
+
+  /// The execution plan to apply the (forward) DFT
+  using dplan = jb::fftw::plan<array_type, frequency_array_type>;
+
+  /// The execution plan to apply the inverse (aka backward) DFT
+  using iplan = jb::fftw::plan<frequency_array_type, output_array_type>;
+
+  /// The type used to store the TDE confidence between two timeseries
+  using confidence_type = jb::fftw::tde_result<array_t, precision_type>;
+
+  /// The type used to store the TDE argmax between two timeseries
+  using argmax_type = jb::fftw::tde_result<array_t, std::size_t>;
+
+  /// The sqr sum of a timeseries
+  using sum2_type = jb::fftw::tde_result<array_t, precision_type>;
+
+  //@}
+
+  /**
+   * Constructs members using the multi array characteristic a and b.
+   *
+   * @param a multi array timeries
+   * @param b multi array timeries
+   *
+   */
+
+  time_delay_estimator_many(array_type const& a, array_type const& b)
```

Comment:
I think we should consider making those inputs non-const if we know the FFTW library is going to change their values.

And we should warn the user in the documentation that the values will be modified.